### PR TITLE
added object.uniformValues for overwriting material.uniforms

### PR DESCRIPTION
**Pull request by arose**
Continuing from #6021. Allow a new property `.uniformValues` in renderable objects that overwrites values in `material.uniforms`.

### CODE REVIEW

File: examples/webgl_interactive_instances_gpu.html [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+
+			</div>
+
+		</div>
+
+		<div id="container"></div>
+
+		<script src="../build/three.min.js"></script>
+
+		<script src="../src/renderers/WebGLRenderer.js"></script>
```
Comment (WestLangley) [inline examples/webgl_interactive_instances_gpu.html]: remove WebGLRenderer source

File: src/renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -4564,6 +4572,25 @@ THREE.WebGLRenderer = function ( parameters ) {
...
+
+		// overwrite uniforms with values from object
+
+		var o_values = object.uniformValues;
+
+		if ( o_values ) {
+
+			for ( var j = 0, jl = o_values.length; j < jl; j ++ ) {
+
+				u = o_values[ j ];
```
Comment (WestLangley) [inline src/renderers/WebGLRenderer.js]: `u` is not declared.

**Comment by WestLangley** [comment]
Nice demo!

If you have a look at `renderer.info` in the console, you will see that the number of programs on the GPU side is not affected by this feature -- the shader program is shared anyway. In light of that, what is the benefit of this PR in your view?

Also, as you change your menu items, `renderer.info` shows the number of geometries increasing. Is there a memory leak of some sort?

**Comment by WestLangley** [comment]
I'm seeing some bugs in which the material color depends on which object is rendered first.

Here is the pattern I am using:

	mesh = new THREE.Mesh( geometry, material );
	mesh.position.set( - 10, 0, 0 );
	scene.add( mesh );

	mesh = new THREE.Mesh( geometry, material );
	mesh.uniformValues = [
		[ "diffuse", new THREE.Color( 0x00ffff ) ]
	];
	mesh.position.set( 10, 0, 0 );
	scene.add( mesh );

**Comment by arose** [comment]
Thanks @WestLangley  for having a look at it!

> remove WebGLRenderer source

I'll keep that in mind for when the PR is ready

> `u` is not declared.

Fixed.

> If you have a look at `renderer.info` in the console, you will see that the number of programs on the GPU side is not affected by this feature -- the shader program is shared anyway. In light of that, what is the benefit of this PR in your view?

It is an optimization to avoid the additional time needed for initializing separate materials for each mesh.

> Also, as you change your menu items, `renderer.info` shows the number of geometries increasing. Is there a memory leak of some sort?

Fixed, I re-created the highlight-box geometry without disposing of it.

> I'm seeing some bugs in which the material color depends on which object is rendered first.

Oh, I had not anticipated that one would use the same material with objects of which some have `.uniformValues` and some not. But, that is a valid usage and is now supported -- the material uniform values are now only temporarily overwritten.

I guess refreshing the uniform values could be made more fine grained, that is, load only those to the GPU that where actually overwritten.

**Comment by WestLangley** [comment]
This is interesting.  I would be nice if, for example, a collection of sprites could share the same sprite sheet, and same material, while having overridden offset/repeat values.

Or, as you said, if all objects could share the same uber material, and have overridden material uniforms -- pushing to the GPU only those overridden values.

**Comment by arose** [comment]
Did an attempt at more fine-grained uniforms GPU upload. This seems to increase rendering speed.

**Comment by mrdoob** [comment]
New approach: #7048

**Pull request closed without merging**